### PR TITLE
Change raptorq decoder

### PR DESCRIPTION
**Pull request by herr-seppia**
### Added

- Add `max_udp_len` configuration parameter
- Add range checks to MTU (between 1296 and 8192)

### Fixed

- Fix raptorQ cache default config
- Fix ObjectTransmissionInformation deserialization

### Changed

- Change the EncodedChunk UUID generation
- Change `raptorq` dependency from `1.6` to `2.0`

### CODE REVIEW

File: src/transport/encoding/raptorq/safe.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+// when it gets stabilized, and this function should be removed.
+// (1) the result is known to not overflow u32 from elsewhere;
+// (2) `denom` is known to not be `0` from elsewhere.
+// TODO this is definitely not always the case! Let's do something about it.
+fn int_div_ceil(num: u64, denom: u64) -> u32 {
+    if num % denom == 0 {
+        (num / denom) as u32
+    } else {
+        (num / denom + 1) as u32
+    }
```
Comment (fed-franz) [inline src/transport/encoding/raptorq/safe.rs:49]: FYI, this can be simplified as

```suggestion
((num + denom - 1) / denom) as u32
```

Feel free to pick your favorite version

**Review by fed-franz — CHANGES_REQUESTED** [review]

File: src/transport.rs [hunk 2]
```diff
@@ -100,8 +99,9 @@ impl WireNetwork {
             }
 
             dec_chan_tx
-                .send((bytes[0..len].to_vec(), remote_address))
-                .await
+                .try_send((bytes[0..len].to_vec(), remote_address))
+                // .send((bytes[0..len].to_vec(), remote_address))
+                // .await
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/transport.rs:104]: Yes

File: src/transport/encoding/raptorq.rs [hunk 3]
```diff
@@ -64,36 +65,40 @@ impl<'a> ChunkedPayload<'a> {
...
         // corrupted transmission info can be received.
-        // If the corrupted info is part of the first received chunk, no message
-        // can ever be decoded.
-        hasher.update(transmission_info);
-        hasher.finalize().into()
+        // If the corrupted info is part of the first received chunk, no
+        // message // can ever be decoded.
+        //
+        // ** UPDATE:
+        // Hashing has been removed in order to increase the decoding
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/transport/encoding/raptorq.rs:99]: I'm sure
I did it (wrongly) previously in order to detect corrupted udp

However, udp corruption is already done at lower OSI level, and if the udp is passed to the application level is guaranteed to be corrected

File: src/transport/encoding/raptorq.rs [hunk 4]
```diff
@@ -163,4 +170,110 @@ mod tests {
         );
         Ok(())
     }
+
+    #[test]
+    fn test_encode_raptorq_junk() -> Result<()> {
+        #[cfg(not(debug_assertions))]
+        let mut data = vec![0; 3_000_000];
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/transport/encoding/raptorq.rs:177]: No it's not
The following declaration set the size to test with if debug is set

File: src/transport/encoding/raptorq.rs [hunk 5]
```diff
@@ -163,4 +170,110 @@ mod tests {
...
+        messge.marshal_binary(&mut writer)?;
+        c = writer.into_inner()?;
+        let mut bytes = vec![];
+        c.seek(std::io::SeekFrom::Start(0))?;
+        c.read_to_end(&mut bytes)?;
+        for i in 44..bytes.len() {
+            bytes[i] = rand::Rng::gen(&mut rand::thread_rng());
+        }
+        let c = Cursor::new(bytes);
+        // c.seek(std::io::SeekFrom::Start(0))?;
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/transport/encoding/raptorq.rs:274]: Yes

File: src/transport/encoding/raptorq.rs [hunk 6]
```diff
@@ -163,4 +170,110 @@ mod tests {
...
+            &TransportDecoder::default_configuration(),
+        );
+        let junks_messages = 100;
+        println!("start spamming with {junks_messages} junk messages");
+        let mut decoded = None;
+        for _ in 0..junks_messages {
+            let mut data = data.clone();
+            for i in 0..data.len() {
+                data[i] = rand::Rng::gen(&mut rand::thread_rng());
+            }
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/transport/encoding/raptorq.rs:212]: To satisfy the borrow checker, since the ownership of the data changes

File: src/transport/encoding/raptorq.rs [hunk 7]
```diff
@@ -163,4 +170,110 @@ mod tests {
...
+        let mut data = vec![0; 100_000];
+
+        for i in 0..data.len() {
+            data[i] = rand::Rng::gen(&mut rand::thread_rng());
+        }
+        let peer = PeerNode::generate("192.168.0.1:666", 0)?;
+        let header = peer.to_header();
+        let payload = BroadcastPayload {
+            height: 255,
+            gossip_frame: data.clone(),
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/transport/encoding/raptorq.rs:189]: I don't get it

File: src/transport/encoding/raptorq/encoder.rs [hunk 8]
```diff
@@ -13,6 +13,10 @@ use crate::transport::Encoder;
 
 const DEFAULT_MIN_REPAIR_PACKETS_PER_BLOCK: u32 = 5;
 const DEFAULT_MTU: u16 = 1300;
+
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/transport/encoding/raptorq/encoder.rs:16]: It's not needs to be multiple of 1024 (indeed 1296 it's not)

File: src/transport/encoding/raptorq/encoder.rs [hunk 9]
```diff
@@ -46,7 +50,17 @@ impl Configurable for RaptorQEncoder {
...
+        let mut conf = *conf;
+        let mtu = conf.mtu;
+        if mtu > MAX_MTU {
+            tracing::warn!("MTU={mtu} too high, changing to {DEFAULT_MTU}");
+            conf.mtu = DEFAULT_MTU;
+        }
+        if mtu < MIN_MTU {
+            tracing::warn!("MTU={mtu} too low, changing to {DEFAULT_MTU}");
+            conf.mtu = DEFAULT_MTU;
+        }
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/transport/encoding/raptorq/encoder.rs:62]: I would prefer to panic instead

File: src/transport/encoding/raptorq.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (fed-franz) [reply src/transport/encoding/raptorq.rs:212]: I mean we don't need to copy the outer `data`, since we're declaring a new `data` variable and fill it with random bytes...

File: src/transport/encoding/raptorq.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (fed-franz) [reply src/transport/encoding/raptorq.rs:189]: I mean that if we remove the cloning at line 209 (where I left the other comment), this clone might become unnecessary.
If line 209 remains unchanged, you can ignore this comment

File: src/transport/encoding/raptorq/safe.rs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+use raptorq::ObjectTransmissionInformation;
+
+use super::{encoder::MAX_MTU, TRANSMISSION_INFO_SIZE};
+
+// This should eventually become <https://doc.rust-lang.org/std/primitive.u64.html#method.div_ceil>
+// when it gets stabilized, and this function should be removed.
+// (1) the result is known to not overflow u32 from elsewhere;
+// (2) `denom` is known to not be `0` from elsewhere.
+// TODO this is definitely not always the case! Let's do something about it.
```
[the remark this answers is not in the document]
Comment (fed-franz) [reply src/transport/encoding/raptorq/safe.rs:43]: I see

File: src/transport/encoding/raptorq/safe.rs (hunk 1 above)
Comment (fed-franz) [reply src/transport/encoding/raptorq/safe.rs:49]: If you change the code you don't have to keep the comments :)

Comment (herr-seppia) [reply src/transport/encoding/raptorq/safe.rs:49]: I would not change the code

Comment (ureeves) [reply src/transport/encoding/raptorq/safe.rs:49]: Its funny because the implementation that @fed-franz suggests is so well known. I understand not wanting to change a file you copied, but its pretty weird the raptor peeps don't know it.

File: src/transport/encoding/raptorq/decoder.rs [hunk 11]
```diff
@@ -13,17 +13,19 @@ use raptorq::{Decoder as ExtDecoder, EncodingPacket};
...
-const DEFAULT_CACHE_TTL_SECS: u64 = 60;
-const DEFAULT_CACHE_PRUNE_EVERY_SECS: u64 = 60 * 5;
+const DEFAULT_CACHE_TTL: Duration = Duration::from_secs(60);
+const DEFAULT_CACHE_PRUNE_EVERY: Duration = Duration::from_secs(30);
+
+const DEFAULT_MAX_UDP_LEN: u64 = 10 * 1_024 * 1_024;
 
 pub struct RaptorQDecoder {
-    cache: HashMap<[u8; 32], CacheStatus>,
+    cache: BTreeMap<[u8; UID_SIZE + TRANSMISSION_INFO_SIZE], CacheStatus>,
```
Comment (ureeves) [inline src/transport/encoding/raptorq/decoder.rs:28]: Why the change to a `BTreeMap`? Do you need ordering?

**Review by ureeves — APPROVED** [review]
Nice. A small question though

File: src/transport/encoding/raptorq/decoder.rs (hunk 11 above)
Comment (herr-seppia) [reply src/transport/encoding/raptorq/decoder.rs:28]: It's part of an audit finding.

Basically it's related to the `retain` function and its performance

_"In the current implementation, this operation takes O(capacity) time instead of O(len) because it internally visits empty buckets too."_ ([src](https://doc.rust-lang.org/std/collections/struct.HashMap.html#performance-7))

**Review by fed-franz — APPROVED** [review]
